FrameworkSystem/DB/ProxyDB.py

```diff
@@ -600,18 +682,87 @@ def __getPUSProxy(self, userDN, userGroup, requiredLifetime, requestedVOMSAttr=N
+    result = ProxyProviderFactory().getProxyProvider(proxyProvider)
+    if not result['OK']:
+      return result
+    pp = result['Value']
+    result = pp.getProxy({"DN": userDN})
+    if not result['OK']:
+      return result
+    proxyStr = result['Value']
+    chain = X509Chain()
+    result = chain.loadProxyFromString(proxyStr)
+    if not result['OK']:
+      return result
+    result = chain.getRemainingSecs()
+    if not result['OK']:
+      return result
+    remainingSecs = result['Value']
+    result = self.__storeProxy(userDN, chain, proxyProvider)
+    if result['OK']:
+      return S_OK({'proxy': proxyStr, 'chain': chain, 'remainingSecs': remainingSecs})
+    return result
+
+  def __getProxyFromProxyProviders(self, userDN, userGroup, requiredLifeTime):
+    """ Generate new proxy from exist clean proxy or from proxy provider
+        for use with userDN in the userGroup
+
+        :param basestring userDN: user DN
+        :param basestring userGroup: required group name
+        :param int requiredLifeTime: required proxy live time in a seconds
+
+        :return: S_OK(tuple)/S_ERROR() -- tuple contain proxy as string and remainig seconds
+    """
+    result = Registry.getGroupsForDN(userDN)
+    if not result['OK']:
+      return S_ERROR('Cannot generate proxy: %s' % result['Message'])
+    if userGroup not in result['Value']:
+      return S_ERROR('Cannot generate proxy: Invalid group %s for user' % userGroup)
+
+    result = Registry.getProxyProvidersForDN(userDN)
+    if result['OK']:
+      for proxyProvider in result['Value']:
+        result = self.__getPemAndTimeLeft(userDN, userGroup, proxyProvider=proxyProvider or 'Certificate')
+        if result['OK'] and (not requiredLifeTime or result['Value'][1] > requiredLifeTime):
+          return result
+        result = self.__generateProxyFromProxyProvider(userDN, proxyProvider)
+        if result['OK']:
+          chain = result['Value']['chain']
+          remainingSecs = result['Value']['remainingSecs']
+          result = chain.generateProxyToString(remainingSecs, diracGroup=userGroup, rfc=True)
+          if result['OK']:
+            return S_OK((result['Value'], remainingSecs))
+    return S_ERROR('Cannot generate proxy%s' %
+                   (result.get('Message') and ': ' + result.get('Message') or ''))
+
   def getProxy(self, userDN, userGroup, requiredLifeTime=False):
     """ Get proxy string from the Proxy Repository for use with userDN
         in the userGroup
-    """
 
+        :param basestring userDN: user DN
+        :param basestring userGroup: required DIRAC group
+        :param int requiredLifeTime: required proxy live time in a seconds
+
+        :return: S_OK(tuple)/S_ERROR() -- tuple with proxy as chain and proxy live time in a seconds
+    """
+    # WARN: this block not need if DIRAC setup use <user name>/DNProperties section in configuration
```

Comment:
I do not understand this comment

Reply:
DIRAC user can have multiple DNs/certificates that provided by some of multiple proxy providers. To match DN with provider, to collect this infomation in CS was added new sections "DNProperties/<some DN>" where described proxy providers that can generate proxy for current DN/certificate for example
If this section will be in every user, provider will be found from this section and additional logic not needed